compiler/fir/resolve/src/org/jetbrains/kotlin/fir/resolve/dfa/FirDataFlowAnalyzer.kt

```diff
@@ -500,20 +530,35 @@ abstract class FirDataFlowAnalyzer<FLOW : Flow>(
         }
 
         logicSystem.approveOperationStatement(flow, predicate).forEach { effect ->
-            flow.addImplication((expressionVariable eq true) implies effect)
-            flow.addImplication((expressionVariable eq false) implies effect.invert())
+            operationStatementTester.invoke(expressionVariable eq true) {
+                flow.addImplication((expressionVariable eq true) implies effect)
+            }
+            operationStatementTester.invoke(expressionVariable eq false) {
+                flow.addImplication((expressionVariable eq false) implies effect.invert())
+            }
         }
 
-        flow.addImplication((expressionVariable eq isEq) implies (operandVariable eq null))
-        flow.addImplication((expressionVariable notEq isEq) implies (operandVariable notEq null))
+        operationStatementTester.invoke(expressionVariable eq isEq) {
+            flow.addImplication((expressionVariable eq isEq) implies (operandVariable eq null))
+        }
+        operationStatementTester.invoke(expressionVariable notEq isEq) {
+            flow.addImplication((expressionVariable notEq isEq) implies (operandVariable notEq null))
+        }
 
-        if (operandVariable is RealVariable) {
-            flow.addImplication((expressionVariable eq isEq) implies (operandVariable typeNotEq any))
-            flow.addImplication((expressionVariable notEq isEq) implies (operandVariable typeEq any))
+        if (operandVariable.isReal()) {
+            operationStatementTester.invoke(expressionVariable eq isEq) {
+                flow.addImplication((expressionVariable eq isEq) implies (operandVariable typeNotEq any))
+            }
+            operationStatementTester.invoke(expressionVariable notEq isEq) {
+                flow.addImplication((expressionVariable notEq isEq) implies (operandVariable typeEq any))
+            }
 
-//            TODO: design do we need casts to Nothing?
-//            flow.addImplication((expressionVariable eq !isEq) implies (operandVariable typeEq nullableNothing))
-//            flow.addImplication((expressionVariable notEq !isEq) implies (operandVariable typeNotEq nullableNothing))
+            operationStatementTester.invoke(expressionVariable eq !isEq) {
+                flow.addImplication((expressionVariable eq !isEq) implies (operandVariable typeNotEq nullableNothing))
```

Comment:
If we want to differentiate `== null` and `is Nothing?`, I think one option is, here, we can use a different cone type with a special type attribute, say, `NullableNothingFromConstantNull` to indicate this `Nothing?` came from `null`, not `is Nothing?`.